tests/integration/test_file_formats.py

```diff
@@ -118,3 +119,31 @@ def test_validate_dataset_type_mismatch(tmp_path, extension):
 
     with pytest.raises(ValidationError, match=errors[extension]):
         validate_dataset(filename, column_specs_2)
+
+
+@pytest.mark.parametrize("extension", list(FILE_FORMATS.keys()))
+def test_validate_dataset_category_mismatch(tmp_path, extension):
+    filename = tmp_path / f"dataset{extension}"
+    column_specs_1 = {
+        "patient_id": ColumnSpec(int),
+        "category": ColumnSpec(str, categories=("a", "b")),
+    }
+    results = [
+        (123, "a"),
+        (456, "b"),
+    ]
+    write_dataset(filename, results, column_specs_1)
+
+    column_specs_2 = {
+        "patient_id": ColumnSpec(int),
+        "category": ColumnSpec(str, categories=("x", "y")),
+    }
+
+    errors = {
+        ".arrow": "Unexpected categories in column 'category'",
```

Comment:
Can we get pyarrow to give something a little more helpful here?

Reply:
The full error message does include the list of categories it was expecting and the list of categories it found. I just didn't bother matching on it here as it made the test harder to read and it felt like enough to ensure that we were triggering the error we expected to trigger.